sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/SpringSdkMessageCodec.scala

```diff
@@ -30,10 +32,59 @@ private[springsdk] object SpringSdkMessageCodec extends MessageCodec {
    * In the Spring SDK, output data are encoded to Json.
    */
   override def encodeScala(value: Any): ScalaPbAny =
-    ScalaPbAny.fromJavaProto(encodeJava(value))
+    ScalaPbAny.fromJavaProto(JsonSupport.encodeJson(value, findTypeHint(value)))
 
   override def encodeJava(value: Any): JavaPbAny =
-    JsonSupport.encodeJson(value)
+    JsonSupport.encodeJson(value, findTypeHint(value))
+
+  private def findTypeHint(value: Any): String = {
+
+    // TODO: cache results, otherwise we'll run scanning + reflection on each encoding
+    def annotatedParents(clz: Class[_], listOfParents: Seq[Class[_]]): Seq[Class[_]] = {
+
+      def hasJsonSubTypes(clz: Class[_]) =
+        clz != null && clz.getAnnotation(classOf[JsonSubTypes]) != null
+
+      val acc =
+        if (hasJsonSubTypes(clz)) listOfParents :+ clz
+        else listOfParents
+
+      val directParents = clz.getSuperclass +: clz.getInterfaces
+
+      directParents.foldLeft(acc) { case (acc, clz) =>
+        if (clz == null) acc // happens when we reach the bottom, ie: Object.getSuperclass == null
+        else annotatedParents(clz, acc)
+      }
+    }
+
+    // straightforward case: class is annotated with JsonTypeName
+    if (value.getClass.getAnnotation(classOf[JsonTypeName]) != null) {
+      value.getClass.getAnnotation(classOf[JsonTypeName]).value()
+    } else {
+      // otherwise needs to scan hierarchy until we find JsonSubTypes annotations
+      // in a parent class or trait
+      val parents = annotatedParents(value.getClass, Seq.empty)
+      if (parents.isEmpty)
+        value.getClass.getName
```

Comment:
I haven't followed what you have discussed for the type hint, but I'm not sure we should use class names.
* It's a source of many of the security problems with Jackson.
* It's volatile for refactoring changes.
* It's not interoperable with non-java clients.

Reply:
Maybe some more context...

We have `typeUrl` using the FQCN all the time. Any state serialized to proto any will have it. Events as well. 

In principle, it will never be a problem for state (famous last words). The reason is that we don't use this information to deserialize anything. It's there because we need to fill in the `typeUrl`. 

When we deserialize the json to some java type, we know already, based on the the user code, which type it should be. User can move their classes around and it will still work. As long as the class and json aligns. 

The `typeUrl` became an issue for the events because now we need this information to be able to dispatch the events to the right methods inside the Spring SDK. See https://github.com/lightbend/kalix/issues/7512

Now we have all interest to provide a stable `typeUrl`.